### ISSUE TITLE
[Feature request] Partial least squares regression

### ISSUE THREAD

**Issue by Nyholm** [markrogoyski/math-php#384]
I've looked at this library and I think Im correct when saying that Partial Least Squares (PLS) regression is missing. It is similar to the principal component regression that is implemented. 

I need this feature and I was hoping that this project would accept a PR. Im not too confident on the math, so if someone know they can do this quickly, then I would be super happy. If not I'll make sure to struggle my way forward =)

See https://en.wikipedia.org/wiki/Partial_least_squares_regression

Question to the maintainer: Will you accept a PR with this?

**Comment by Beakerboy** [comment]
We’re happy for any improvements! Make sure you fully test the functions against the results from an independent source. For example, if a textbook has examples with solutions, or you can use R, octave, or numpy wiTh the same data.

**Comment by Beakerboy** [comment]
Using the process from this document

https://personal.utdallas.edu/~herve/Abdi-PLS-pretty.pdf

I was able to use Excel to manually get the same results as the following R code:
```
library(mixOmics)

sample_data = as.data.frame(matrix(c(7, 4, 10, 16, 13, 7, 3, 5, 7, 3, 13, 14, 12, 11, 10, 7, 7, 5, 3, 3), nrow = 5, ncol = 4))
colnames(sample_data) <- c("input1","input2","input3","input4")
pred = as.data.frame(matrix(c(14, 10, 8, 2, 6, 7, 7, 5, 4, 2, 8, 6, 5, 7, 4), nrow = 5, ncol = 3))
sample_data
colnames(pred) <- c("output1", "output2", "output3")
pred
pls.model <- pls(sample_data, pred, ncomp=3)
pls.model$loadings
```

So that seems like a workable algorithm

**Comment by Nyholm** [comment]
Thank you. This is much helpful. Do you mind share the excel doc too?

**Comment by Beakerboy** [comment]
I deleted the excel spreadsheet. PHP-ish psudocode follows. I have no idea how one actually uses these matrices in practice, so i would not be comfortable coding this class up myself with meaningful tests without a lot more research on my part
    
```php
// Prepare Data, can use similar process as in PCA class
$E = standardize columns of X
$F = Standardized columns of Y
$PC = 0
while (E !== Null_Matrix) {
  $u_new = random_vector
  do {
    $u = $u_new;
    $w = Eᵀ * u / ‖Eᵀ * u‖; //(unit vector in direction of Eᵀ * u); 
    $t = E * w / ‖E * w‖;
    $c = Fᵀ * t / ‖Fᵀ * t‖;
    $u_new=F * c;
  } while ($unew !== $u);
  $p = $E->transpose()->multiply($t);
  $b = $t->transpose()->multiply($u); 
  $T[++$PC] = $t  //move $t to the $PC column of $T
  // Do the same with w, u, p, and c
  $E = $E->subtract($t->multiply($p->transpose()));
  $F = $F->subtract($t->multiply($c->transpose())->multiply($b));
}

**Comment by Beakerboy** [comment]
From the R documentation

“Many ... formulations can be found in literature. It has been shown, for instance, that only one of X and Y needs to be deflated; alternatively, one can directly deflate the crossproduct matrix S (as is done in SIMPLS, for example). Moreover, there are many equivalent ways of scaling. In the example above, the scores t have been normalised, but one can also choose to introduce normalisation at another point in the algorithm. Unfortu- nately, this can make it difficult to directly compare the scores and loadings of different PLSR implementations.“

**Comment by Beakerboy** [comment]
Some code sources below:
PLS2 looks like it is the most straightforward, and is what is described in the wikipedia article, however, most examples online appear to use the mvr function. It looks like several PLSR algorithms use SVD, which has not been integrated into mathPHP yet. PLS1/2 do not rely on SVD,
  
https://rdrr.io/cran/chemometrics/src/R/pls2_nipals.R
https://rdrr.io/cran/pls/src/R/mvr.R

**Comment by Beakerboy** [comment]
As an update #385 is a pull request for PLS in MathPHP. I would appreciate any feedback if you are able to review it to see if it gives you your expected results, and that it has the features you would expect.

**Comment by Beakerboy** [comment]
This can be closed

**Comment by markrogoyski** [comment]
PLS is available in the latest release [v2.2.0](https://github.com/markrogoyski/math-php/releases/tag/v2.2.0)
